src/Parallel/NodeLock.cpp

```diff
@@ -10,69 +10,72 @@
 
 namespace Parallel {
 
-NodeLock::NodeLock() noexcept { lock_ = CmiCreateLock(); }
+NodeLock::NodeLock() noexcept
+    : lock_(std::make_unique<CmiNodeLock>(CmiCreateLock())) {}
 
-NodeLock::NodeLock(NodeLock&& moved_lock) noexcept {
-  moved_lock.destroy();
-  lock_ = CmiCreateLock();
+NodeLock::NodeLock(NodeLock&& moved_lock) noexcept
+    : lock_(std::move(moved_lock.lock_)) {
+  moved_lock.lock_ = nullptr;
 }
 
 NodeLock& NodeLock::operator=(NodeLock&& moved_lock) noexcept {
-  moved_lock.destroy();
-  lock_ = CmiCreateLock();
+  lock_ = std::move(moved_lock.lock_);
```

Comment:
So thinking about this more, I don't know if it's safe to move a lock a) at all really, and b) when it's unlocked. Definitely not while it's unlocked, I think. I'm not sure what the right answer is, to be honest. Do we use move semantics for locks anywhere? I guess when creating the DataBox?

I think the ways to resolve this are:
- use a `std::atomic<std::shared_ptr<CmiLock>>` so that we can have atomic access to the lock. (seems a bit silly to have an atomic lock, but if we want to be able to move the lock I think that's semi-necessary)
- make accessing a lock at the same time as it is being `move`d undefined behavior, user bug, no diagnostic, good luck folks.
Whether the second is acceptable depends on how much (and when) we will be moving locks.

Reply:
I assume you mean it's not safe to move a locked lock?   Items in the DataBox must be movable for `db::create` and `db::create_from`.   So maybe have the moves and pup (and destructor?) fail if the lock is locked?

Reply:
No, you get the issue even if it's unlocked. Say A decides to do a `move` and half a cycle later B starts locking, then the pointer might be null between when B started locking and A moved (or something much more confusing). Basically, there is a race condition here that I have no idea what would happen if you get it wrong. It'll be a disaster for sure :P While this case is _extremely_ unlikely, debugging it would be probably one of the worst debug experiences ever. I think the solution is using `std::atomic_shared_ptr` and then doing the move with `std::atomic_exchange`. I think since the exchange can make it a `nullptr` it's still a bit spooky.

Reply:
I guess, I'm not sure a moved-from NodeLock can safely be a `nullptr` without another atomic variable to check... I'm scared of this whole thing, but I don't know of the right solution, honestly.

Reply:
Is this different from a normal data race?  The move operation is not thread safe, so it is the job of the programmer to ensure that it does not occur while another thread is accessing the object.

Reply:
I think it's true that we can use `std::unique_ptr` to make the move semantics here a bit easier. Would it work to just make `NodeLock` un-movable and change the tags so that we store a `std::unique_ptr<NodeLock>` in the databox and in the algorithm member variable?

We'd presumably continue to deal with migration in the same kind of solution we currently have.